clean-abap/sub-sections/ModernABAPLanguageElements.md

```diff
@@ -0,0 +1,630 @@
+                      ELSE 7 ).
+```
+
+Old style:
+
+```ABAP
+DATA value TYPE i.
+IF status = open.
+  value = 1.
+ELSEIF status = blocked.
+  value = 3.
+ELSE.
+  value = 7.
+ENDIF.
+```
+
+> Alternatively you may use the [function `xsdbool( )`](../CleanABAP.md#use-xsdbool-to-set-boolean-variables)
+
+#### Case distinction
+
+Evaluate case distinction with the `SWITCH #( )` operator
+
+```ABAP
+DATA(status) = SWITCH #( scrum_status
+    WHEN scrum_status_open THEN status_waiting
+    WHEN scrum_status_in_process THEN status_busy
+    WHEN scrum_status_blocked THEN status_alarm
+    WHEN scrum_status_done THEN status_ok ).
+```
+
+Old style:
+
+```ABAP
+DATA status TYPE status_enum.
+CASE scrum_status.
+  WHEN scrum_status_open.
+    status = status_waiting.
+  WHEN scrum_status_in_process.
+    status = status_busy.
+  WHEN scrum_status_blocked.
+    status = status_alarm.
+  WHEN scrum_status_done.
+    status = status_ok.
+ENDCASE.
+```
+
+#### Case distinctions of reference types class and interface
+
+Switch on a reference types class and interface using the `CASE` extension `TYPE OF`.
+
+```ABAP
+CASE type of account.
+  WHEN TYPE bank_account INTO DATA(bank_account).
+    " some processing ...
+  WHEN OTHERS.
+    " some processing ...
+ENDCASE.
+```
+
+In a condition e.g. in an `IF` statement the `IS INSTANCE OF` operator can be used.
+
+```ABAP
+IF account IS INSTANCE OF bank_account.
+  " some processing ...
+ENDIF.
+```
+
+### Conversions
+
+#### Cast data references
+
+Cast reference types to other reference types use the `CAST #( )` operator.
+
+```ABAP
+DATA(my_account) = CAST account( NEW bank_account( ) ).
+```
+
+Old style:
+
+```ABAP
+DATA my_account TYPE REF TO account.
+CREATE OBJECT my_account TYPE bank_account.
+```
+
+or
+
+```ABAP
+DATA my_account TYPE REF TO account.
+DATA my_bank_account TYPE REF TO bank_account.
+CREATE OBJECT my_bank_account.
+my_account ?= bank_account.
+```
+
+#### Create data references
+
+Create data references to structures and tables with the operator `REF #( )`.
+
+```ABAP
+DATA accounts TYPE accounts_table.
+import_accounts_references( REF #( accounts ) ).
+```
+
+Old style:
+
+```ABAP
+DATA accounts TYPE accounts_table.
+DATA accounts_reference TYPE REF TO accounts_type.
+GET REFERENCE OF accounts INTO accounts_reference.
+import_accounts_references( accounts_reference ) ).
+```
+
+#### Convert data types
+
+Use the operator `CONV #( )` to convert data types and save temporary variables.
+
+```ABAP
+method_takes_string( CONV #( a_char ) ).
+```
+
+Old style:
+
+```ABAP
+DATA a_string TYPE string.
+a_string = a_char.
+method_takes_string( a_string ).
+```
+
+#### Copy fields with matching names
+
+Copy fields with matching names from one data type to another with `corresponding #( )`.
+
+```ABAP
+target_structure = CORRESPONDING #( source_structure ).
+```
+
+Old style:
+
+```ABAP
+MOVE-CORRESPONDING source_structure TO target_structure.
+```
+
+> The two differ slightly in behavior.
```

Comment:
I think _slightly_ is a bit of an understatement in some cases. Maybe add a comment or a link?
```abap
DATA(saved_msg) = CORRESPONDING symsg( sy ).
do_stuff_that_changes_msg_vars( ).
MOVE-CORRESPONDING saved_msg TO sy. " Restore msg variables
sy = CORRESPONDING #( saved_msg ). " Restore msg variables and clear all other fields (!)
```

Reply:
True. "Slightly" may be slightly misleading. :-) @thomham added a better formulation.